src/qt/walletcontroller.cpp

```diff
@@ -124,3 +119,24 @@ void WalletController::removeAndDeleteWallet(WalletModel* wallet_model)
     // CWallet shared pointer.
     delete wallet_model;
 }
+
+
+OpenWalletActivity::OpenWalletActivity(WalletController* wallet_controller, const std::string& name)
+    : m_wallet_controller(wallet_controller)
+    , m_name(name)
+{}
+
+void OpenWalletActivity::open()
+{
+    std::string error, warning;
+    std::unique_ptr<interfaces::Wallet> wallet = m_wallet_controller->m_node.loadWallet(m_name, error, warning);
+    if (!warning.empty()) {
+        Q_EMIT message(QMessageBox::Warning, QString::fromStdString(warning));
+    }
+    if (wallet) {
+        Q_EMIT opened(m_wallet_controller->getOrCreateWallet(std::move(wallet)));
```

Comment:
In commit "gui: Add OpenWalletActivity" (8847cdaaaeb45c1ddee89f43ac4b8fafb20e5c0d)

@promag, would there there any downsides to deleting this line? It seems redundant and as far as I can tell. Everything seems to works fine without it due to the load wallet event.

I'm trying to implement my suggestion from https://github.com/bitcoin/bitcoin/pull/15153#discussion_r249868084, and I'd really prefer just to have one way to attach a Wallet to the gui, not two different ways.

Reply:
Note that the signal `opened` is connected to the `setCurrentlyWallet` but there is currently a problem there because the wallet view is created after, which needs to be fixed - I was already aware of this.

I don't think that's a good idea - suppose you have 2 windows, you open a wallet on each (because the loading is async and doesn't block the GUI) you should see each wallet on the expected window.

Reply:
Thanks, a multi window case is helpful to think about. I know you are working on fixing sync stuff in the gui, but what I would expect when loading a slow wallet is for the window to show "Loading wallet X..." right after the user selects the wallet, so the user has some immediate feedback. Then when the loadWallet notification is triggered, the GUI would use the Wallet handle passed with the notification to display actual wallet information. There would be no need for the call which the GUI makes to request loading of the wallet to return a wallet handle.

It sounds like this a concern for the future, though. Thanks for answering my immediate question.

Reply:
But how would it know if the notified wallet handle belongs to the wallet opened?

Reply:
Yes, I was just thinking wallet name.